azure-sdk-for-cpp: init at various versions

tobim opened the pull request:
My motivation to have this packaged is so we can enable ABFS support in `arrow-cpp`.

Each API client library is built in a dedicated derivation to minimize closure sizes of dependent packages.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tobim committed:
azure-sdk-for-cpp: init at various versions

cpcloud commented on pkgs/by-name/az/azure-sdk-for-cpp/attestation.nix:
```diff
@@ -0,0 +1,53 @@
+    moveToOutput "share" $dev
+  '';
+
+  doCheck = false;
```
Is there no hope for running tests for any of these?

cpcloud replied:
Looks like testing is [quite an endeavor](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#testing-the-project).

tobim replied:
Thanks for reminding me. I'll give it another try tonight or tomorrow.

tobim replied:
I looked into this a bit more. I don't think investing time in that is worth it.

cpcloud reviewed (approved):
Given the constraints, this seems reasonable.

FliegendeWurst commented on pkgs/by-name/az/azure-sdk-for-cpp/package.nix:
```diff
@@ -0,0 +1,210 @@
+    tables
+  ];
+
+  buildPhase = null;
+  installPhase = ''
+    mkdir $out
+  '';
```
```suggestion
  dontBuild = true;
  installPhase = ''
    runHook preInstall
    mkdir $out
    runHook postInstall
  '';
```

FliegendeWurst commented on pkgs/by-name/az/azure-sdk-for-cpp/package.nix:
```diff
@@ -0,0 +1,210 @@
+  src = fetchFromGitHub {
+    owner = "Azure";
+    repo = "azure-sdk-for-cpp";
+    rev = "azure-identity_1.10.0";
```
```suggestion
    tag = "azure-identity_1.10.0";
```

SuperSandro2000 reviewed (commented):
What is your plan to keep all those version numbers in sync?

ShamrockLee commented:
How could I help with this PR and PR #426611?

Both @osbm's nixpkgs-review results and OfBorg build logs show that azure-sdk-for-cpp.core-amqp is failing to build on Darwin platforms. Should we mark those platforms as unsupported for now?

tobim committed:
azure-sdk-for-cpp: use release tags for each component

osbm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 348202`
Commit: `e6add369a2c406f5da98cb6125f9f15cc37f7a42`

osbm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 348202`
Commit: [`e6add369a2c406f5da98cb6125f9f15cc37f7a42`](https://github.com/NixOS/nixpkgs/commit/e6add369a2c406f5da98cb6125f9f15cc37f7a42) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/e6add369a2c406f5da98cb6125f9f15cc37f7a42..pull/348202/head))
Merge: [`35ad840a3c85ea6a504f6e3360fd97da42012cea`](https://github.com/NixOS/nixpkgs/commit/35ad840a3c85ea6a504f6e3360fd97da42012cea)

Logs: https://github.com/osbm/nixpkgs-review-gha/actions/runs/18935202512


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `x86_64-darwin`
:white_check_mark: *No rebuilds*

---
### `aarch64-darwin`
:white_check_mark: *No rebuilds*

osbm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 348202`
Commit: [`24a0121c59d11d5b2bdd28c5b160dfa43ec3f205`](https://github.com/NixOS/nixpkgs/commit/24a0121c59d11d5b2bdd28c5b160dfa43ec3f205) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/24a0121c59d11d5b2bdd28c5b160dfa43ec3f205..pull/348202/head))
Merge: [`ad3c120f8c8a033eef606d9d39666f974600a018`](https://github.com/NixOS/nixpkgs/commit/ad3c120f8c8a033eef606d9d39666f974600a018)

Logs: https://github.com/osbm/nixpkgs-review-gha/actions/runs/19093916351


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

osbm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 348202`
Commit: [`97c73c46a7784feb45033dcfc5855b7ce9defa93`](https://github.com/NixOS/nixpkgs/commit/97c73c46a7784feb45033dcfc5855b7ce9defa93) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/97c73c46a7784feb45033dcfc5855b7ce9defa93..pull/348202/head))
Merge: [`6b1b50b0226c2fa9d6b2cfaab5ad02a126ab75ff`](https://github.com/NixOS/nixpkgs/commit/6b1b50b0226c2fa9d6b2cfaab5ad02a126ab75ff)

Logs: https://github.com/osbm/nixpkgs-review-gha/actions/runs/19099371259


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-sdk-for-cpp</li>
  </ul>
</details>

tobim committed:
azure-sdk-for-cpp: expose scope as passthru.pkgs

tobim commented on pkgs/by-name/az/azure-sdk-for-cpp/package.nix:
Thanks for these suggestions.

I opted to keep the dedicated scope, as these packages are strongly related to each other. I'm not using `makeScopeWithSplicing'`, because all packages in the set are only used with `{propagated,}buildInputs`.

ShamrockLee reviewed (dismissed):
Just noticed that you use `propagatedBuildInputs`. Sorry for this late review.

ShamrockLee commented on pkgs/by-name/az/azure-sdk-for-cpp/package.nix:
```diff
@@ -0,0 +1,96 @@
+stdenvNoCC.mkDerivation (finalAttrs: {
+  name = "azure-sdk-for-cpp";
+
+  propagatedBuildInputs = [
```
How about we make `azure-sdk-for-cpp` the scope, and exposes a set `azure-sdk-for-cpp.sdkPackages` for people to append to their package's `buildInputs`?

tobim replied:
I think turning the `azure-sdk-for-cpp` into the scope is the right idea, but `sdkPackages` isn't needed after all, as it only leads to false build dependencies.

tobim replied:
Well, by-name packages must be derivations according to the rules. I don't know whether to revert that or move it out of by-name.

infinisil replied:
Should be moved out of `pkgs/by-name` yeah.

tobim replied:
Moved to `/pkgs/development/libraries/azure-sdk-for-cpp` now.

tobim committed:
azure-sdk-for-cpp: Drop wrapping derivation

It's cleaner to make the SDK a dedicated scope and force call sites
to list the modules they actually need.
This makes the top-level package significantly simpler.

tobim committed:
azure-sdk-for-cpp: move to all-packages

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/core-amqp.nix:
```diff
@@ -0,0 +1,78 @@
+    c-shared-utility
+    macro-utils-c
+    umock-c
+  ];
```
`propagatedBuildInputs` should be avoided in most cases and should only be used to workaround build system's limitation.

A typical use case is when a package provides some CMake/pkg-config configuration which, when used by other packages, expects certain indirect dependencies to be present. Is that the reason why you are using it?

```suggestion
  buildInputs = [
    core
    c-shared-utility
    macro-utils-c
    umock-c
  ];
```

tobim replied:
> Is that the reason why you are using it?

Precisely. For example `messaging-eventhubs` doesn't list any of those as direct dependencies, but fails to configure if they are not present transitively.

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/core-amqp.nix:
```diff
@@ -0,0 +1,78 @@
+  ];
+
+  env = {
+    AZURE_SDK_DISABLE_AUTO_VCPKG = 1;
```
(Not a blocker.)

Is the opposite `AZURE_SDK_DISABLE_AUTO_VCPKG=0` or `AZURE_SDK_DISABLE_AUTO_VCPKG=`? If the latter, consider

```nix
{
  env = {
    AZURE_SDK_DISABLE_AUTO_VCPKG = true;
  };
}
```

ShamrockLee commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1513,6 +1513,8 @@ with pkgs;
 
   azure-cli-extensions = recurseIntoAttrs azure-cli.extensions;
 
+  azure-sdk-for-cpp = callPackage ../development/libraries/azure-sdk-for-cpp { };
```
We need Hydra to build it and make it available in the binary cache.

```suggestion
  azure-sdk-for-cpp = recurseIntoAttrs (callPackage ../development/libraries/azure-sdk-for-cpp { });
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/core-amqp.nix:
```diff
@@ -0,0 +1,78 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
This line also moves `share/azure-core-amqp-cpp/copyright` to `$dev`. If a dependent package's output content only references `azure-sdk-for-cpp.core-amqp.out` but not `azure-sdk-for-cpp.core-amqp.dev`, the `copyright` file won't be downloaded to the users' machine, which violates the MIT license this package is distributed under.

```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-core-amqp-cpp/copyright" "$out"
```

ShamrockLee reviewed (dismissed):
The `copyright` file location is a blocker, and the rest are not.

tobim commented on pkgs/development/libraries/azure-sdk-for-cpp/default.nix:
```diff
@@ -0,0 +1,66 @@
+    }
+    // lib.optionalAttrs stdenv.hostPlatform.isDarwin {
+      inherit apple-sdk;
+    }
```
I think I don't understand this request. What dependent package definitions do you mean? The only one that uses it here is `c-shared-utility`, and that also checks `stdenv.hostPlatform.isDarwin`.

tobim replied:
Thanks for the explanation. I moved it up now.

tobim committed:
azure-sdk-for-cpp: make it build on hydra

Co-authored-by: Yueh-Shun Li <shamrocklee@posteo.net>

tobim committed:
azure-sdk-for-cpp: keep the copyright in "out"

Co-authored-by: Yueh-Shun Li <shamrocklee@posteo.net>

tobim committed:
azure-sdk-for-cpp: add a note about disabled checks

tobim commented on pkgs/development/libraries/azure-sdk-for-cpp/core-tracing-opentelemetry.nix:
```diff
@@ -0,0 +1,70 @@
+    moveToOutput "share" $dev
+  '';
+
+  doCheck = false;
```
Added a note above that line in `core.nix`.

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/data-tables.nix:
```diff
@@ -0,0 +1,75 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-data-tables-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/identity.nix:
```diff
@@ -0,0 +1,71 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-identity-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/messaging-eventhubs-checkpointstore-blob.nix:
```diff
@@ -0,0 +1,75 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-messaging-eventhubs-checkpointstore-blob-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/security-attestation.nix:
```diff
@@ -0,0 +1,73 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-security-attestation-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/security-keyvault-keys.nix:
```diff
@@ -0,0 +1,69 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/azure-security-keyvault-keys-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/core-amqp.nix:
```diff
@@ -0,0 +1,80 @@
+
+  postInstall = ''
+    moveToOutput "share" "$dev"
+    moveToOutput "share/azure-core-amqp-cpp/copyright" "$out"
```
```suggestion
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/security-keyvault-secrets.nix:
```diff
@@ -0,0 +1,69 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/storage-common.nix:
```diff
@@ -0,0 +1,75 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/storage-files-datalake.nix:
```diff
@@ -0,0 +1,73 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/storage-files-shares.nix:
```diff
@@ -0,0 +1,69 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee commented on pkgs/development/libraries/azure-sdk-for-cpp/storage-queues.nix:
```diff
@@ -0,0 +1,69 @@
+  ];
+
+  postInstall = ''
+    moveToOutput "share" $dev
```
```suggestion
    moveToOutput "share" "$dev"
    moveToOutput "share/$(basename "$sourceRoot")-cpp/copyright" "$out"
```

ShamrockLee reviewed (commented):
Here are all the `moveToOutput` suggestions.

ShamrockLee reviewed (approved):
LGTM!

I think this one is good enough to merge. We can add enhancements like splicing and version checks in subsequent PRs.

Thank you for pushing it!
